### PR TITLE
 MAINT: polynomial: Add an N-d vander implementation used under the hood of the vander2d and vander3d functions

### PR DESCRIPTION
**Pull request by eric-wieser**
~~First commit is gh-14352, which this refactor depends on. Marking as WIP until that is merged.~~

### CODE REVIEW

File: numpy/polynomial/polyutils.py [hunk 1]
```diff
@@ -415,51 +416,87 @@ def mapdomain(x, old, new):
...
+            "Expected {} dimensions of sample points, got {}".format(n_dims, len(points)))
+    if n_dims != len(degrees):
+        raise ValueError(
+            "Expected {} dimensions of degrees, got {}".format(n_dims, len(degrees)))
+    if n_dims == 0:
+        raise ValueError("Unable to guess a dtype or shape when no points are given")
+
+    # convert to the same shape and type
+    points = tuple(
+        np.array(tuple(points), copy=False) + 0.0
```
Comment (charris) [inline numpy/polynomial/polyutils.py]: This doesn't guarantee float64 or complex128 as documented. See `as_series` above. That function could maybe use some improvement.

Comment (charris) [reply numpy/polynomial/polyutils.py]: If the shapes don't match, an object array is produced.

File: numpy/polynomial/polyutils.py [hunk 2]
```diff
@@ -415,51 +416,87 @@ def mapdomain(x, old, new):
...
+    )
+    # we checked this wasn't empty already, so no `initial` needed
+    return functools.reduce(operator.mul, (
+        # slicing inserts a dimension per degree
+        vander_f(p, deg)[(...,) + _nth_slice(i, n_dims)]
```
Comment (charris) [inline numpy/polynomial/polyutils.py]: It might be clearer to just put this in a `for` loop, I doubt there would be much difference in speed.

**Review by charris — CHANGES_REQUESTED** [review]
The shape check of the points and the conversion to a common type needs some fixing

File: numpy/polynomial/polyutils.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/polynomial/polyutils.py]: This is the same as what was always present in the separate 2d and 3d functions though - so I'd argue this is out of scope

File: numpy/polynomial/polyutils.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/polynomial/polyutils.py]: Putting this in a for loop  ends up as something messy like:
```python
result = None
for ... in zip(...):
    part = vander_f(p, deg)[(...,) + _nth_slice(i, n_dims)
    if result is not None:
        result += part
    else:
        result = part
assert result is not None
```

Comment (charris) [reply numpy/polynomial/polyutils.py]: That doesn't look right. Something like
```
aslice = (slice(None),)
result = vander_fs[0](points[0], degrees[0])
for i in range(1, n_dims):
    aslice = (None,) + aslice
    result = result[..., None] * vander_fs[i](points[i], degs[i])[(...,) + aslice]
return result
```
Returning the result of a function call with a complicated set of arguments just does not read well and I don't think worth the use of reduce. There are lots of ways to break up the steps.

Comment (eric-wieser) [reply numpy/polynomial/polyutils.py]: IMO incrementally slicing `result` is more confusing - your loop replaces a simple reduce operation  (multiply) with a complex one (slice and multiply)

How about I split my version over multiple lines:
```python
# produce the vander-monde matrix for each dimension, placing each of the trailing axes in a unique location
vander_arrays = (
    vander_f(p, deg)[(...,) + _nth_slice(i, n_dims)]
    for i, (vander_f, p, deg) in enumerate(zip(vander_fs, points, degrees))
)

return functools.reduce(operator.mul, vander_arrays)
```

Comment (charris) [reply numpy/polynomial/polyutils.py]: I like the split versions better, with a slight (0+) preference for the second. Maybe indent the closing parenthesis.

Note that that incremental slicing has fewer operations and one less function to track down, but don't care much one way or the other. It is like an inductive proof :)

Comment (eric-wieser) [reply numpy/polynomial/polyutils.py]: I'm not a huge fan of induction unless its necessary.

Either way - mind reviewing gh-14352 before I update this one?

Comment (eric-wieser) [reply numpy/polynomial/polyutils.py]: > Note that that incremental slicing has fewer operations

You have two slices per access, whereas I only have one. Admittedly your slices are cheaper to construct though. I doubt it matters, and if it did `_nth_slice` could become a lookup table.

**Comment by eric-wieser** [comment]
Ok to punt changing the coercion logic as requested in the hidden chain at https://github.com/numpy/numpy/pull/14353#discussion_r319695550? I've added a remark in the commit message saying that the old behavior has been preserved.

**Comment by eric-wieser** [comment]
Anything else needed here @charris? Would you prefer not to put this in at all?

**Review by charris — APPROVED** [review]

**Comment by charris** [comment]
Yep, let's get this in, then finish up the fitting.

**Pull request merged**